# COVIDvu - US County Data Comparison (2 Sources) <img src='../resources/American-flag.png' align = 'right'>

Ticket #0364 requires a comparison of two sets of data for US COVID-19 Reporting. This is data from CSBS.org and
NY Times. The following information is needed:

1. Number of counties reported in each data set for the time when you analyze them
2. Statistical differences in the total (which is greater?)
3. Per county comparison
4. Per state comparison

---
# County Totals

### Clone the NY Times Data Repo

COVIDvu is poised to use the CSBS data (API available) for US reporting of confirmed cases, however another source was found
where the NY Times is gathering similar data and providing that via Github repo and CSV formatted files. This data is not updated but cummulative (i.e. ever-growing). The CSBS data is JSON and maintains updated data (current date).

In [8]:
%%bash
if [ -d "./covid-19-data" ]
then
  echo "Directory ./covid-19-data exists."
else
  git clone https://github.com/nytimes/covid-19-data.git
fi

Directory ./covid-19-data exists.


### Clone the CSBS Data API Repo (Reference)

In [9]:
%%bash
if [ -d "./coronavirus-tracker-api" ]
then
  echo "Directory ./coronavirus-tracker-api exists."
else
  git clone https://github.com/ExpDev07/coronavirus-tracker-api.git
fi

Directory ./coronavirus-tracker-api exists.


---
# Strategy

To obtain the data required for our analysis, we should get the snapshots (same date) into similar data structures (JSON). 
The NYTimes data is by day and ever-growing. The CSBS API data is near real-time. These differences make it difficult to 
provide an accurate analysis until both sets of data take the same form. First-take on structure would include county, state
(province), cases, and deaths. Date meanings may be very different if not even comparable. Max-cases reported has more meaning.

### DataFrame NYTimes (All Rows)

This is the entire dataset with the fips column removed for clarity. Ultimately the date column will be removed as it
is not really needed for our comparison of near-realtime data.

In [58]:
import json 
import numpy as np
import pandas as pd 
from pandas import json_normalize #package for flattening json in pandas df

nytimes = pd.read_csv ('covid-19-data/us-counties.csv', index_col=False)
nytimes[['date','county','state','cases','deaths']]

date      county       state  cases  deaths
0      2020-01-21   Snohomish  Washington      1       0
1      2020-01-22   Snohomish  Washington      1       0
2      2020-01-23   Snohomish  Washington      1       0
3      2020-01-24        Cook    Illinois      1       0
4      2020-01-24   Snohomish  Washington      1       0
...           ...         ...         ...    ...     ...
15831  2020-03-27        Park     Wyoming      1       0
15832  2020-03-27    Sheridan     Wyoming      5       0
15833  2020-03-27  Sweetwater     Wyoming      1       0
15834  2020-03-27       Teton     Wyoming     12       0
15835  2020-03-27    Washakie     Wyoming      1       0

[15836 rows x 5 columns]

### DataFrame CSBS (All Rows)

Before we can render our CSBS dataframe, we need to pull the data locally via API into a JSON file. This is done here:

In [59]:
%%bash
export COUNTY_CASES_US_ENDPOINT="https://coronavirus-tracker-api.herokuapp.com/v2/locations?source=csbs"
echo "request county data from CSBS API at $COUNTY_CASES_US_ENDPOINT"

curl -s "$COUNTY_CASES_US_ENDPOINT" | jq -M '.locations' > "countyraid-csbs.json" 

request county data from CSBS API at https://coronavirus-tracker-api.herokuapp.com/v2/locations?source=csbs


In [60]:
#load json object
with open('countyraid-csbs.json') as f:
    d = json.load(f)

#lets put the data into a pandas df
csbs = json_normalize(d)
csbs.rename(columns={'last_updated':'date','province':'state', 'latest.confirmed':'cases', 'latest.deaths':'deaths'}, inplace=True)
csbs['date'] = pd.to_datetime(csbs['date']).dt.date
csbs.drop_duplicates(subset=['county','state','cases'], keep='last', inplace=True)

csbs[['date','county','state','cases','deaths']]


date       county     state  cases  deaths
0     2020-03-29     New York  New York  32308     678
1     2020-03-29  Westchester  New York   7875      10
2     2020-03-29       Nassau  New York   5537      34
3     2020-03-29      Suffolk  New York   4138      37
4     2020-03-29     Rockland  New York   1896       8
...          ...          ...       ...    ...     ...
1938  2020-03-29     Seminole  Oklahoma      1       0
1939  2020-03-29        Texas  Oklahoma      1       0
1940  2020-03-29   Montgomery    Kansas      3       0
1941  2020-03-29     Stafford    Kansas      1       0
1942  2020-03-29      Stevens    Kansas      1       0

[1943 rows x 5 columns]

### DataFrame NYTimes (Unique Counties)

In [61]:
nytimes.sort_values(by=['state', 'county', 'date'], inplace = True)
nytimes.drop_duplicates(subset=['county','state'], keep='last', inplace=True)
nytimes[['date','county','state','cases','deaths']]

date      county    state  cases  deaths
14046  2020-03-27     Autauga  Alabama      6       0
14047  2020-03-27     Baldwin  Alabama      5       0
14048  2020-03-27      Blount  Alabama      5       0
14049  2020-03-27     Bullock  Alabama      2       0
14050  2020-03-27      Butler  Alabama      1       0
...           ...         ...      ...    ...     ...
15831  2020-03-27        Park  Wyoming      1       0
15832  2020-03-27    Sheridan  Wyoming      5       0
15833  2020-03-27  Sweetwater  Wyoming      1       0
15834  2020-03-27       Teton  Wyoming     12       0
15835  2020-03-27    Washakie  Wyoming      1       0

[1797 rows x 5 columns]

---
# Per County Comparison 

In [62]:
nytimes['county'].isin(csbs['county']).value_counts()

True     1724
False      73
Name: county, dtype: int64

In [69]:
c_compare = pd.merge(nytimes, csbs, on=['state','county'], how='inner')
c_compare.sort_values(by=['state','county'], inplace = True)
c_compare.rename(columns={'cases_x':'cases_nytimes','cases_y':'cases_csbs'}, inplace=True)
c_compare['match?'] = np.where(c_compare['cases_nytimes'] == c_compare['cases_csbs'], 'True', 'False')
c_compare['positive?'] = np.where(c_compare['cases_nytimes'] > c_compare['cases_csbs'], 'True', 'False')
c_compare['negative?'] = np.where(c_compare['cases_nytimes'] < c_compare['cases_csbs'], 'True', 'False')
c_compare[['county','state','cases_nytimes','cases_csbs','match?','positive?','negative?']]
#c_compare.to_csv('countyraid-county-compare.csv')

county    state  cases_nytimes  cases_csbs match? positive?  \
0        Autauga  Alabama              6           6   True     False   
1        Baldwin  Alabama              5          14  False     False   
2         Blount  Alabama              5           5   True     False   
3        Bullock  Alabama              2           3  False     False   
4         Butler  Alabama              1           1   True     False   
...          ...      ...            ...         ...    ...       ...   
1710        Park  Wyoming              1           1   True     False   
1711    Sheridan  Wyoming              5           6  False     False   
1712  Sweetwater  Wyoming              1           1   True     False   
1713       Teton  Wyoming             12          14  False     False   
1714    Washakie  Wyoming              1           1   True     False   

     negative?  
0        False  
1         True  
2        False  
3         True  
4        False  
...        ...  
1710     False  
1711      True  
1712     False  
1713      True  
1714     False  

[1715 rows x 7 columns]

### Percentage that CSBS has more reported cases

In [74]:
s_compare['negative?'].value_counts(normalize=True) * 100

True     96.078431
False     3.921569
Name: negative?, dtype: float64

---
# Per State Comparison

In [24]:
nytimes['state'].isin(csbs['state']).value_counts()

True     1796
False       1
Name: state, dtype: int64

In [72]:
s_compare = pd.merge(nytimes, csbs, on=['state','county'], how='inner')
s_compare.sort_values(by=['state','county','cases_y'], inplace = True)
s_compare.groupby('state')['cases_y'].sum()
s_compare.sort_values(by=['state','cases_y'], inplace = True)
s_compare.drop_duplicates(subset=['state'], keep='last', inplace=True)
s_compare.rename(columns={'cases_x':'cases_nytimes','cases_y':'cases_csbs'}, inplace=True)
#s_compare.size

s_compare['match?'] = np.where(s_compare['cases_nytimes'] == s_compare['cases_csbs'], 'True', 'False')
s_compare['positive?'] = np.where(s_compare['cases_nytimes'] > s_compare['cases_csbs'], 'True', 'False')
s_compare['negative?'] = np.where(s_compare['cases_nytimes'] < s_compare['cases_csbs'], 'True', 'False')
s_compare['total_csbs'] = s_compare['cases_csbs'].sum()
s_compare['total_nytimes'] = s_compare['cases_nytimes'].sum()

s_compare[['state','cases_nytimes','cases_csbs','match?','positive?','negative?', 'total_nytimes','total_csbs']]
#s_compare.to_csv('countyraid-state-compare.csv')

state  cases_nytimes  cases_csbs match? positive?  \
26                 Alabama            179         214  False     False   
53                  Alaska             43          49  False     False   
59                 Arizona            399         545  False     False   
99                Arkansas             89          92  False     False   
124             California           1465        1818  False     False   
167               Colorado            312         367  False     False   
199            Connecticut            752         908  False     False   
208               Delaware            106         130  False     False   
210   District of Columbia            304         342  False     False   
238                Florida            869        1121  False     False   
299                Georgia            347         407  False     False   
364                 Hawaii             87         108  False     False   
370                  Idaho             99         114  False     False   
390               Illinois           2239        2613  False     False   
462                Indiana            484         676  False     False   
516                   Iowa             58          65  False     False   
553                 Kansas             66         101  False     False   
589               Kentucky             76          97  False     False   
645              Louisiana           1170        1350  False     False   
670                  Maine             92         142  False     False   
692               Maryland            208         301  False     False   
711          Massachusetts            631         843  False     False   
756               Michigan            824        1018  False     False   
783              Minnesota            141         171  False     False   
819            Mississippi             63          71  False     False   
923               Missouri            247         313  False     False   
931                Montana             42          58  False     False   
951               Nebraska             60          57  False      True   
963                 Nevada            443         443   True     False   
975          New Hampshire             75          86  False     False   
979             New Jersey           1505        2169  False     False   
999             New Mexico             82          93  False     False   
1065              New York           7187        7875  False     False   
1103        North Carolina            218         315  False     False   
1128          North Dakota             25          28  False     False   
1153                  Ohio            330         370  False     False   
1225              Oklahoma             94         120  False     False   
1261                Oregon            122         154  False     False   
1301          Pennsylvania            530         865  False     False   
1315          Rhode Island             42         156  False     False   
1323        South Carolina             92         108  False     False   
1371          South Dakota             18          25  False     False   
1388             Tennessee            312         376  False     False   
1455                 Texas            367         488  False     False   
1540                  Utah            212         276  False     False   
1549               Vermont             90         117  False     False   
1568              Virginia            124         188  False     False   
1620            Washington           1830        2077  False     False   
1652         West Virginia             23          29  False     False   
1683             Wisconsin            468         541  False     False   
1705               Wyoming             17          23  False     False   

     negative?  total_nytimes  total_csbs  
26        True          25658       31013  
53        True          25658       31013  
59        True          25658       3101

### Percentage that CSBS has more reported cases

In [73]:
s_compare['negative?'].value_counts(normalize=True) * 100

True     96.078431
False     3.921569
Name: negative?, dtype: float64

---
# Experiment Results

---
## Conclusions for No. 1

### Number of counties reported in each data set for the time when you analyze them

- NYTimes reporting 1797 US Counties as of 2020-03-27
- CSBS reporting 1875 US Counties as of 2020-03-27

NOTE: Both datasets may have independent cities (e.g. Virginia) or boroughs (e.g. Alaska) that are bordered by county but reporting as an entity that does not reside in a county proper.

## Conclusions for No. 2

The CSBS data contains reporting county entities but these entities are not reported from a county seat but they still
remain a child of the US state. These extra entities may not be official counties but they still require representation. This does not, however prove that the CSBS data is more accurate but it does provide more reported COVID-19 cases confirmed.

## Conclusions for No. 3

During the per county comparison of COVID-19 reported cases, the CSBS data was updated by mistake of running the notebook again against the CSBS API data. This data is more up to date than the data being collected by NYTimes. The NYTimes data is stuck at 2020-03-27, however the CSBS data is being updated every day (current 2020-03-29). 

While the data was at matched by date, the confirmed cases were higher with the CSBS data than the NYTimes collection.

## Conclusions for No. 4

While the per state comparison was matched by at last_updated date (2020-03-27) the overall state confirmed cases of COVID-19
were higher in the CSBS data as well.